src/Workspaces/Core/Portable/Workspace/Workspace.cs

```diff
@@ -196,13 +197,56 @@ protected Solution SetCurrentSolution(Solution solution)
             while (true)
             {
                 var newSolution = solution.WithNewWorkspace(this, currentSolution.WorkspaceVersion + 1);
-                var replacedSolution = Interlocked.CompareExchange(ref _latestSolution, newSolution, currentSolution);
-                if (replacedSolution == currentSolution)
+                var oldSolution = Interlocked.CompareExchange(ref _latestSolution, newSolution, currentSolution);
+                if (oldSolution == currentSolution)
                 {
                     return newSolution;
                 }
 
-                currentSolution = replacedSolution;
+                currentSolution = oldSolution;
+            }
+        }
+
+        /// <summary>
+        /// Applies specified transformation to <see cref="CurrentSolution"/>, updates <see cref="CurrentSolution"/> to the new value and raises a workspace change event of the specified kind.
+        /// </summary>
+        /// <param name="transformation">Solution transformation.</param>
+        /// <param name="kind">The kind of workspace change event to raise.</param>
+        /// <param name="projectId">The id of the project updated by <paramref name="transformation"/> to be passed to the workspace change event.</param>
+        /// <param name="documentId">The id of the document updated by <paramref name="transformation"/> to be passed to the workspace change event.</param>
+        /// <returns>True if <see cref="CurrentSolution"/> was set to the transformed solution, false if the transformation did not change the solution.</returns>
+        internal bool SetCurrentSolution(Func<Solution, Solution> transformation, WorkspaceChangeKind kind, ProjectId? projectId = null, DocumentId? documentId = null)
```

Comment:
Should this have some validation if `kind` is something that requires a projectId or documentId but they didn't pass one?

Reply:
(I see this hasn't been done before, so I guess this is fine, although I wonder what other bugs are lurking...)